### PR TITLE
X-Ray Localized Sampling (2/2)

### PR DESCRIPTION
**Pull request by marcomagdy**
This patch completes X-Ray support in Envoy (minus centralized sampling which is coming next).
However with this patch, users can use AWS X-Ray with a JSON manifest file and get traces showing in their AWS console.

Risk Level: Low
Testing: unit testing & manual
Docs Changes: none
Release Notes: none

Signed-off-by: Marco Magdy <mmagdy@gmail.com>

### CODE REVIEW

**Comment by alyssawilk** [comment]
@htuch based on codeowners I'd like to assign @lavignes to first pass, but looks like they need an invite to the org.

Drive by comment: the CODEOWNERS file should include a maintainer.  I'd suggest @jmarantz since he and @htuch did the initial commit.  If he doesn't object, mind adding his name to CODEOWNERS as part of this PR?

**Comment by jmarantz** [comment]
Should we add @lavignes to the org and have him do a first pass before I take a look?

**Comment by htuch** [comment]
Org invite sent to @lavignes

**Comment by lavignes** [comment]
@alyssawilk @htuch Thanks. You can assign to me. :)

File: source/extensions/tracers/xray/daemon.proto [hunk 1]
```diff
@@ -0,0 +1,24 @@
+syntax = "proto3";
+
```
Comment (lavignes) [inline source/extensions/tracers/xray/daemon.proto]: To clarify, this was added as a means to serialize the messages to the daemon into JSON?

Perhaps a comment here to explain that?

File: source/extensions/tracers/xray/daemon.proto [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+package envoy.tracers.xray.daemon;
+
+// see https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+message Segment {
```
Comment (lavignes) [inline source/extensions/tracers/xray/daemon.proto]: Do you think we should add some validations? ie. Mark some strings as non-empty, etc. The segment document spec you link actually lays out maximum string lengths and such fairly well. Even if this is just for documentation purposes it's nice to have the reference here.

Or perhaps add comments inline as to what these represent and expected values.

File: source/extensions/tracers/xray/daemon_broker.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <memory>
```
Comment (lavignes) [inline source/extensions/tracers/xray/daemon_broker.h]: I don't think these headers are needed here

File: source/extensions/tracers/xray/tracer.cc [hunk 4]
```diff
@@ -1,37 +1,134 @@
...
+// For example, 1-58406cf0-a006649127e371903a2de979.
+// This includes:
+//
+// - The version number, that is, 1.
+// - The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+// - A 96-bit unique identifier in 24 hexadecimal digits.
+//
+// For more details see:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+std::string generateTraceId(Envoy::SystemTime point_in_time) {
```
Comment (lavignes) [inline source/extensions/tracers/xray/tracer.cc]: Isn't it more conventional to put this in a utility class?

File: source/extensions/tracers/xray/tracer.h [hunk 5]
```diff
@@ -18,14 +21,18 @@ namespace Extensions {
...
-static const char XRayTraceHeader[] = "x-amzn-trace-id";
+constexpr auto XRayTraceHeader = "x-amzn-trace-id";
 
 class Span : public Tracing::Span {
 public:
   /**
-   * Construct a new X-Ray Span.
+   * Creates a new Span.
+   *
+   * @param broker Facilitates communication with the X-Ray daemon.
```
Comment (lavignes) [inline source/extensions/tracers/xray/tracer.h]: nit: params are in the wrong order

File: source/extensions/tracers/xray/tracer.h [hunk 6]
```diff
@@ -18,14 +21,18 @@ namespace Extensions {
...
+   * @param time_source A time source to create Span IDs using the monotonic clock.
    */
-  Span(TimeSource& time_source) : time_source_(time_source) {}
+  Span(Envoy::TimeSource& time_source, DaemonBroker* broker)
```
Comment (lavignes) [inline source/extensions/tracers/xray/tracer.h]: Why not pass/store the broker as a reference? It should never be null.

**Comment by marcomagdy** [comment]
@jmarantz actually, I've had @lavignes and @abaptiste review this internally before I submitted it publicly.
So while their input is always welcome, please don't let that hold you back from reviewing this when you get a chance. Thank you :)

File: source/extensions/tracers/xray/tracer.cc (hunk 4 above)
Comment (lavignes) [reply source/extensions/tracers/xray/tracer.cc]: Also I don't think this function has test coverage. Shouldn't we inject the RNG?

Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.cc]: >Isn't it more conventional to put this in a utility class?

But is it useful to put it in a utility class?
it's used only for x-ray trace IDs (happens only in this file).
Also, because it's not in a class, I'm able to mark it as "internal linkage".

> Also I don't think this function has test coverage. 

It _does_ have coverage by means of testing the functions calling it.

>Shouldn't we inject the RNG?

I'm not sure what the downsides are of not injecting it.

File: source/extensions/tracers/xray/tracer.h (hunk 5 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.h]: will fix it. Thanks.

File: source/extensions/tracers/xray/tracer.h (hunk 6 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.h]: Yes. That's a left over from an earlier refactoring. Will change it. Thanks.

File: source/extensions/tracers/xray/daemon_broker.h (hunk 3 above)
Comment (lavignes) [reply source/extensions/tracers/xray/daemon_broker.h]: Whoops. :)

File: source/extensions/tracers/xray/tracer.cc (hunk 4 above)
Comment (lavignes) [reply source/extensions/tracers/xray/tracer.cc]: I guess I didn't see a test that explicitly tests the output to validate that it returns the 24 character hyphen-delimited output in the correct format. Im more worried about making sure there it a regression test specifically for that.

File: source/extensions/tracers/xray/tracer.cc [hunk 7]
```diff
@@ -1,37 +1,138 @@
...
+
+} // namespace
+
+void Span::setId(uint64_t id) { id_ = Hex::uint64ToHex(id); }
+
 void Span::finishSpan() {
-  end_time_ = time_source_.systemTime();
-  // TODO(marcomagdy): serialize this span and send it to X-Ray Daemon
+  using std::chrono::time_point_cast;
+  using namespace envoy::tracers::xray;
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.cc]: see above about `using namespace`.

File: source/extensions/tracers/xray/tracer.cc [hunk 8]
```diff
@@ -1,37 +1,138 @@
...
+    throw new EnvoyException("Failed to serialize span to JSON.");
+  }
+
+  broker_.send(json);
 }
 
 void Span::injectContext(Http::HeaderMap& request_headers) {
-  const std::string xray_header_value =
-      fmt::format("root={};parent={};sampled=1", traceId(), parent_segment_id_);
+  const std::string& xray_header_value =
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.cc]: Curious: why did you add the `&` here: this is not a common pattern for this usage AFAICT. I realize why it works, but also note that what was there before did not make extra copies either.

File: source/extensions/tracers/xray/tracer.cc [hunk 9]
```diff
@@ -44,17 +145,59 @@ Tracing::SpanPtr Tracer::startSpan(const std::string& span_name, const std::stri
...
+  constexpr auto SpanContentLength = "content_length";
+  constexpr auto SpanMethod = "method";
+  constexpr auto SpanStatus = "status";
+  constexpr auto SpanUserAgent = "user_agent";
+  constexpr auto SpanUrl = "url";
+  constexpr auto HttpUrl = "http.url";
+  constexpr auto HttpMethod = "http.method";
+  constexpr auto HttpStatusCode = "http.status_code";
+  constexpr auto HttpUserAgent = "user_agent";
+  constexpr auto HttpResponseSize = "response_size";
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.cc]: it seems surprising to me that these are local to a particular function. Shouldn't they be declared in a place that the consumers of the tag can reference?

Or, if these are meant for sending to a non-OSS system, maybe a comment pointing to the doc enumerating the spans would be good.

Also I would probably not use `auto` here; I'd probably use:
```
  constexpr char SpanContentLength[] = "content_length";
```
etc

As I'm not even sure what the `auto` will resolve to in this case.

File: source/extensions/tracers/xray/tracer.h [hunk 10]
```diff
@@ -18,14 +21,18 @@ namespace Extensions {
 namespace Tracers {
 namespace XRay {
 
-static const char XRayTraceHeader[] = "x-amzn-trace-id";
+constexpr auto XRayTraceHeader = "x-amzn-trace-id";
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.h]: I am more confident the earlier form is optimal.

File: source/extensions/tracers/xray/tracer.h [hunk 11]
```diff
@@ -18,14 +21,18 @@ namespace Extensions {
...
 public:
   /**
-   * Construct a new X-Ray Span.
+   * Creates a new Span.
+   *
+   * @param time_source A time source to create Span IDs using the monotonic clock.
+   * @param broker Facilitates communication with the X-Ray daemon.
    */
-  Span(TimeSource& time_source) : time_source_(time_source) {}
+  Span(Envoy::TimeSource& time_source, DaemonBroker& broker)
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.h]: s/Envoy:://

File: source/extensions/tracers/xray/tracer.h [hunk 12]
```diff
@@ -107,62 +110,77 @@ class Span : public Tracing::Span {
...
-      : segment_name_(segment_name), time_source_(time_source) {
-    UNREFERENCED_PARAMETER(time_source_);
-  }
+  Tracer(absl::string_view segment_name, DaemonBrokerPtr daemon_broker, TimeSource& time_source)
+      : segment_name_(segment_name), daemon_broker_(std::move(daemon_broker)),
+        time_source_(time_source) {}
 
   /**
-   * Starts a tracing span for X-Ray
+   * Start a tracing span for X-Ray
```
Comment (jmarantz) [inline source/extensions/tracers/xray/tracer.h]: The prior spelling is slightly more common I think, but up to you.

File: source/common/common/hex.cc [hunk 13]
```diff
@@ -60,4 +60,16 @@ std::string Hex::uint64ToHex(uint64_t value) {
 
   return encode(&data[0], data.size());
 }
+
+std::string Hex::uint32ToHex(uint32_t value) {
+  std::array<uint8_t, 4> data;
+
+  // This is explicitly done for performance reasons
```
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/common/common/hex.cc]: [I have microbenchmarked it](http://quick-bench.com/bpWXIuvhBaKZiS2cAaSwWLTckl8). And it's about 1500 faster than `stringstream`.

Also, this happens to be the same algorithm used by the function above it for the 64-bit version.

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize X-Ray header to JSON.");
+  }
+  return json;
+}
+
+} // namespace
+
+DaemonBrokerImpl::DaemonBrokerImpl(const std::string& daemon_endpoint) {
+  using Network::Utility;
```
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: Did you mean to say "compliant"? if not, I need clarification, please.

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  auto& logger = Logger::Registry::getLog(Logger::Id::tracing);
+
+  const auto address = Utility::parseInternetAddressAndPort(daemon_endpoint, false /*v6only*/);
+  io_handle_ = address->socket(Network::Address::SocketType::Datagram);
+  if (io_handle_->fd() == -1) {
+    ENVOY_LOG_TO_LOGGER(logger, error, "Failed to acquire UDP socket to X-Ray daemon at - {}",
+                        daemon_endpoint);
```
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: good catch. fixed.

Definitely not throwing :) I dislike exceptions so much.

File: source/extensions/tracers/xray/tracer.cc (hunk 4 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.cc]: I've added a test.

File: source/extensions/tracers/xray/tracer.cc (hunk 8 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.cc]: The `&` saves a move unless (N)RVO kicks in, then it's just as efficient.

File: source/extensions/tracers/xray/tracer.cc (hunk 9 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.cc]: I've added a comment pointing to the documentation.

>  `constexpr char SpanContentLength[] = "content_length";`

clang-tidy complains "Avoid C-style arrays" when I use that notation, which I think is absurd in this context. But I also hate the warning showing up all over my files.

> As I'm not even sure what the auto will resolve to in this case.

`constexpr auto SpanContentLength ...;` is equivalent to `constexpr char* SpanContentLength ...` which is equivalent to `const char * const SpanContentLength ...`

File: source/extensions/tracers/xray/tracer.h (hunk 10 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.h]: How so? https://godbolt.org/z/rHUhf9

File: source/extensions/tracers/xray/tracer.h (hunk 12 above)
Comment (marcomagdy) [reply source/extensions/tracers/xray/tracer.h]: I want it to be consistent across the API.
I'm ok to change them all. WDYT?

File: source/extensions/tracers/xray/tracer.h (hunk 10 above)
Comment (jmarantz) [reply source/extensions/tracers/xray/tracer.h]: I did not know about godbolt.org. That's awesome. My reasoning about `[]` being the more efficient form may not apply if you used `constexpr`. godbolt.org shows it's the exact same assembly emitted.

I was thinking that the `*` form provided an assignable pointer in addition to the static array, but evidently not so in your form. Thanks!

File: source/extensions/tracers/xray/tracer.h (hunk 12 above)
Comment (jmarantz) [reply source/extensions/tracers/xray/tracer.h]: My guess is that a significant percentage of Envoy would consistently use "Starts a tracing span..." and the rest would be random. So if you want to be consistent with Envoy style I'd suggest that.

But as I said it's up to you.

File: source/extensions/tracers/xray/daemon_broker.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/extensions/tracers/xray/daemon_broker.cc]: save status in the object then?

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+  const auto result = address->connect(io_handle_->fd());
+  if (result.rc_ == -1) {
+    ENVOY_LOG_TO_LOGGER(logger, error, "Failed to send X-Ray UDP packet to - {}", daemon_endpoint);
+  }
+}
+
+void DaemonBrokerImpl::send(const std::string& data) const {
+  constexpr auto version = 1;
+  constexpr auto format = "json";
+  const std::string payload = absl::StrCat(createHeader(format, version), "\n", data);
```
Comment (jmarantz) [inline source/extensions/tracers/xray/daemon_broker.cc]: you didn't use the reference form here. I think this instance is by far the more common pattern in Envoy.

File: source/extensions/tracers/xray/tracer.cc (hunk 8 above)
Comment (jmarantz) [reply source/extensions/tracers/xray/tracer.cc]: see above about consistency with using the ref form or not.

File: source/common/common/hex.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/common/hex.cc]: Oh I didn't know about http://quick-bench.com/ either -- this is also very cool. I'm learning lots of great stuff in this review.

Put a comment indicating the perf difference into the code here?

File: source/extensions/tracers/xray/daemon_broker.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: Can you explain your thoughts about adding a status data member in this class?

I'm honestly not sure when it would be the case of failing to acquire a UDP socket, and whether that is something that would persist or might change if we keep retrying...

[the remark this answers is not in the document]
Comment (jmarantz) [reply source/extensions/tracers/xray/daemon_broker.cc]: Yes, if something goes wrong opening the socket, we need to define what the system behavior should be after that.

I'm not sure whether the behavior should be:
 * crash with error message
 * run in degraded state
 * something else

But I don't have a clear idea on what the behavior should be in this case.

[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: Well, right now the behavior is to log the error and retry again the next time.. I think that is acceptable personally, it can also be thought of as running in a degraded state as well. 

Is there a particular reason you are _not_ OK with that approach?

File: source/common/common/hex.cc [hunk 17]
```diff
@@ -58,6 +58,19 @@ std::string Hex::uint64ToHex(uint64_t value) {
...
 
-  return encode(&data[0], data.size());
+  return encode(data.data(), data.size());
+}
+
+std::string Hex::uint32ToHex(uint32_t value) {
+  std::array<uint8_t, 4> data;
+
+  // This is explicitly done for performance reasons
+  // using std::stringstream with std::hex is ~3 orders of magnitude slower
```
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/common/common/hex.cc]: I was going to add the link, but I'm not sure of the longevity of the url and that website SLA when it comes to that. I didn't want to end up with a dead link in the comments, that's all. WDYT?

**Comment by marcomagdy** [comment]
@mattklein123 since there seems to be an ownership dilemma here, can we ship this as alpha? (Note: I don't have permissions to label the PR as alpha)

File: source/common/common/hex.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/common/hex.cc]: OK then I think it might be easiest to just add it to utility_speed_test.cc since we have that already.

File: source/extensions/tracers/xray/daemon_broker.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/extensions/tracers/xray/daemon_broker.cc]: Just paranoia. What happens if it's in a state it never succeeds? Will it spam the log file?

I'm OK with your approach if you just add some comments saying what happens & why that's OK.

[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/tracers/xray/daemon_broker.cc]: Acquiring the socket itself should never fail IIRC so this can be a release assert. The connect below can definitely fail, and that should not be an error case, as it can spam the log. I would recommend actually getting rid of connect entirely and just using sendmsg() directly (see all of the UDP code that has been merged recently). This will avoid an issue like: https://github.com/envoyproxy/envoy/issues/8911 which I'm going to fix this week by removing the connect and moving to sendmsg() :)

[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: Looking at `Network::IoHandle::sendmsg(...)`, is there a way to create a `RawSlice` without copying the data into one of the `Slice` abstractions? seems like a bit of an overkill compared to the current solution (modulo the log spam problem).

[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/tracers/xray/daemon_broker.cc]: https://github.com/envoyproxy/envoy/blob/80cea119e0910e9c41ba543017bfc3ee0cb0909f/source/common/network/utility.h#L303-L305

[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon_broker.cc]: I guess I'd have to ask the same question about `Buffer::Instance`. Is there a way to create one without copying the data into it?

[the remark this answers is not in the document]
Comment (mattklein123) [reply source/extensions/tracers/xray/daemon_broker.cc]: Just use the slice variant and point to the data. It's just a string view. See other examples in the code base.

File: source/common/common/hex.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (marcomagdy) [reply source/common/common/hex.cc]: Talked offline to @jmarantz about this. Adding a test won't be necessary

File: api/envoy/config/trace/v2alpha/xray.proto [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+option java_outer_classname = "XRayTracer";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.trace.v2alpha";
+
+import "envoy/api/v2/core/base.proto";
+
+import "validate/validate.proto";
+
+// [#not-implemented-hide:]
```
Comment (mattklein123) [inline api/envoy/config/trace/v2alpha/xray.proto]: This is now implemented. Can you remove this, make sure the docs render nicely, and also add a release note about this new feature?

File: api/envoy/config/trace/v2alpha/xray.proto [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+// Configuration for AWS X-Ray tracer.
+message Config {
+  // The endpoint of the X-Ray Daemon where the spans will be sent.
+  // This should be set in the form of ip:port.
+  // If this value is not set, the default value of 127.0.0.1:2000 will be used.
+  string daemon_endpoint = 1 [(validate.rules).string = {min_bytes: 1}];
```
Comment (mattklein123) [inline api/envoy/config/trace/v2alpha/xray.proto]: This should use `api.v2.core.Address` (see other examples for stats backends for example).

File: api/envoy/config/trace/v2/trace.proto [hunk 20]
```diff
@@ -46,8 +45,8 @@ message Tracing {
     // - :ref:`DynamicOtConfig <envoy_api_msg_config.trace.v2.DynamicOtConfig>`
     // - :ref:`DatadogConfig <envoy_api_msg_config.trace.v2.DatadogConfig>`
     // - :ref:`OpenCensusConfig <envoy_api_msg_config.trace.v2.OpenCensusConfig>`
-    // [#comment: TODO(marco) when XRay is implemented, uncomment the following; - :ref:`XRayConfig
-    // <envoy_api_msg_config.trace.v2.XRayConfig>`]
+    // [#comment: TODO(marcomagdy) when XRay is implemented, uncomment the following; - :ref:`XRay
```
Comment (mattklein123) [inline api/envoy/config/trace/v2/trace.proto]: I think we can call this support done now so let's uncomment this. We can do future PRs as enhancements.

File: source/extensions/tracers/xray/daemon.proto [hunk 21]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+// The structures in this file are merely for JSON serialization to communicate with the X-Ray
+// daemon. The X-Ray daemon does not speak protobuf.
+// TODO(marcomagdy): is there anyway to prevent protoc from generating types for java/C#/etc.?
```
Comment (mattklein123) [inline source/extensions/tracers/xray/daemon.proto]: What do you mean here? We shouldn't be generating any Java/C#/etc. code for C++.

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  envoy::tracers::xray::daemon::Header header;
+  header.set_format(format);
+  header.set_version(version);
+
+  Protobuf::util::JsonPrintOptions json_options;
+  json_options.preserve_proto_field_names = true;
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(header, &json, json_options);
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize X-Ray header to JSON.");
```
Comment (mattklein123) [inline source/extensions/tracers/xray/daemon_broker.cc]: Can this actually fail? If not I would assert this. If so please add a test for this.

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  return json;
+}
+
+} // namespace
+
+DaemonBrokerImpl::DaemonBrokerImpl(const std::string& daemon_endpoint) {
+  address_ = Network::Utility::parseInternetAddressAndPort(daemon_endpoint, false /*v6only*/);
+  io_handle_ = address_->socket(Network::Address::SocketType::Datagram);
+  RELEASE_ASSERT(
```
Comment (mattklein123) [inline source/extensions/tracers/xray/daemon_broker.cc]: This assert already lives in `socketFromSocketType` and can be removed.

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+  constexpr auto format = "json";
+  const std::string payload = absl::StrCat(createHeader(format, version), "\n", data);
+  Buffer::RawSlice buf;
+  buf.mem_ = const_cast<char*>(payload.data());
+  buf.len_ = payload.length();
+  const auto rc = Network::Utility::writeToSocket(*io_handle_, &buf, 1 /*num_slices*/,
+                                                  nullptr /*local_ip*/, *address_);
+
+  if (rc.rc_ != payload.length()) {
+    ENVOY_LOG_TO_LOGGER(logger, error, "Failed to send trace payload to the X-Ray daemon.");
```
Comment (mattklein123) [inline source/extensions/tracers/xray/daemon_broker.cc]: This will log spam badly. I would make this a debug/trace log and have a stat if you want to have an operational way of looking at this.

File: source/extensions/tracers/xray/tracer.cc [hunk 25]
```diff
@@ -1,37 +1,139 @@
...
+
+  // TODO(marcomagdy): test how expensive this validation is. Might be worth turning off in
+  // optimized builds..
+  MessageUtil::validate(s, ProtobufMessage::getStrictValidationVisitor());
+  Protobuf::util::JsonPrintOptions json_options;
+  json_options.preserve_proto_field_names = true;
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(s, &json, json_options);
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize span to JSON.");
```
Comment (mattklein123) [inline source/extensions/tracers/xray/tracer.cc]: Same comment about failing. Can this fail?

File: source/extensions/tracers/xray/tracer.cc [hunk 26]
```diff
@@ -44,17 +146,61 @@ Tracing::SpanPtr Tracer::startSpan(const std::string& span_name, const std::stri
       span_ptr->setSampled(true);
       break;
     case SamplingDecision::NotSampled:
-      span_ptr->setSampled(false);
+      // should never get here. If the header has Sampled=0 then we never call startSpan().
+      RELEASE_ASSERT(false, "X-Ray startSpan called despite HTTP header having sampled=0");
```
Comment (mattklein123) [inline source/extensions/tracers/xray/tracer.cc]: swap to NOT_REACHED

**Comment by marcomagdy** [comment]
I'm a little at a loss here...
The v3alpha file that is causing erros in the CI is auto generated by the proto_sync file...
I'm not sure what I'm doing wrong at this point...
```
ERROR: /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy_api/envoy/config/trace/v3alpha/BUILD:7:1: no such package '@envoy_api//envoy/config/trace/v2alpha/xray': BUILD file not found in directory 'envoy/config/trace/v2alpha/xray' of external repository @envoy_api. Add a BUILD file to a directory to mark it as a package. and referenced by '@envoy_api//envoy/config/trace/v3alpha:pkg'
ERROR: Analysis of target '@envoy_api//envoy/config/trace/v3alpha:pkg' failed; build aborted: no such package '@envoy_api//envoy/config/trace/v2alpha/xray': BUILD file not found in directory 'envoy/config/trace/v2alpha/xray' of external repository @envoy_api. Add a BUILD file to a directory to mark it as a package.
```

**Comment by mattklein123** [comment]
Might be a bug in the transform tool, not sure. @htuch @lizan can you PTAL at the error above?

**Comment by marcomagdy** [comment]
Looks like I had the x-ray proto file with a namespace that does not map 1:1 with the folder structure and that threw off the transform tool.

File: api/envoy/config/trace/v2alpha/xray.proto [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+  // The UDP endpoint of the X-Ray Daemon where the spans will be sent.
+  // If this value is not set, the default value of 127.0.0.1:2000 will be used.
+  api.v2.core.SocketAddress daemon_endpoint = 1;
+
+  // The name of the X-Ray segment. By default this will be set to the cluster name.
+  string segment_name = 2;
+
+  // The location of a local custom sampling rules JSON file.
+  // For an example of the sampling rules see:
+  // https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-go-configuration.html#xray-sdk-go-configuration-sampling
```
Comment (mattklein123) [inline api/envoy/config/trace/v2alpha/xray.proto]: nit: Can you turn this into an actual link? See docs for how to do and check rendered docs please.

File: source/extensions/tracers/xray/config.cc [hunk 28]
```diff
@@ -29,8 +33,18 @@ XRayTracerFactory::createHttpTracerTyped(const envoy::config::trace::v2::XRayCon
...
 
-  XRayConfiguration xconfig{proto_config.daemon_endpoint(), proto_config.segment_name(),
-                            sampling_rules_json};
+  RELEASE_ASSERT(proto_config.daemon_endpoint().protocol() ==
+                     api::core::SocketAddress::Protocol::SocketAddress_Protocol_UDP,
+                 "X-Ray daemon endpoint must be a UDP socket address");
+
+  RELEASE_ASSERT(proto_config.daemon_endpoint().port_specifier_case() ==
+                     api::core::SocketAddress::PortSpecifierCase::kPortValue,
+                 "X-Ray daemon port must be specified as number. Not a named port.");
```
Comment (mattklein123) [inline source/extensions/tracers/xray/config.cc]: These are actual config errors and shouldn't crash. Can you turn these into exceptions with appropriate tests?

File: source/extensions/tracers/xray/daemon_broker.cc [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+  const auto status = Protobuf::util::MessageToJsonString(header, &json, json_options);
+  ASSERT(status.ok());
+  return json;
+}
+
+} // namespace
+
+DaemonBrokerImpl::DaemonBrokerImpl(const std::string& daemon_endpoint) {
+  address_ = Network::Utility::parseInternetAddressAndPort(daemon_endpoint, false /*v6only*/);
+  io_handle_ = address_->socket(Network::Address::SocketType::Datagram);
```
Comment (mattklein123) [inline source/extensions/tracers/xray/daemon_broker.cc]: nit: these can be initialized in the initializer list and then made const members.

**Comment by marcomagdy** [comment]
One thing I'm not happy with is the location where X-Ray tracer shows up..
This is an artifact of moving x-ray to its own proto file (I think).
LMK if that's acceptable, if not, please advise on where it should show up..

envoy-docs

**Comment by marcomagdy** [comment]
Also, I checked the coverage report and the only thing missing are the lines which send the data over UDP. In the future i’ll look into adding an integration test.

File: docs/root/api-v2/bootstrap/bootstrap.rst [hunk 30]
```diff
@@ -11,3 +11,4 @@ Bootstrap
   ../config/overload/v2alpha/overload.proto
   ../config/ratelimit/v2/rls.proto
   ../config/trace/v2/trace.proto
+  ../config/trace/v2alpha/xray.proto
```
Comment (mattklein123) [inline docs/root/api-v2/bootstrap/bootstrap.rst]: What I would recommend doing here in a follow up is just making a new directory/TOC for tracing, and then put these two RST in that with its own TOC. This will make the new docs look better. Later we can split out all the other tracers into their own files also. This is not a big deal so feel free to do this in your next change.

File: source/extensions/tracers/xray/tracer.cc [hunk 31]
```diff
@@ -1,37 +1,136 @@
 #include "extensions/tracers/xray/tracer.h"
 
+#include <sys/socket.h>
```
Comment (mattklein123) [inline source/extensions/tracers/xray/tracer.cc]: nit: I don't think this is needed? You can delete in your next change.

File: source/extensions/tracers/xray/daemon.proto [hunk 32]
```diff
@@ -0,0 +1,28 @@
+syntax = "proto3";
+
+// The structures in this file are merely for JSON serialization to communicate with the X-Ray
+// daemon. The X-Ray daemon does not speak protobuf.
+package envoy.tracers.xray.daemon;
```
Comment (htuch) [inline source/extensions/tracers/xray/daemon.proto]: @marcomagdy How come this is in the `envoy.tracers` namespace? This seems likely to cause some confusion with API protos which share the namespace. I hit this while developing some tooling around the APIs that noticed this type.

Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon.proto]: Happy to change it. I wasn't super sure what namespace it should go under. I couldn't find many examples of proto files in non-api locations.

should it be `envoy.extensions.tracers.xray.daemon`?

Comment (htuch) [reply source/extensions/tracers/xray/daemon.proto]: If it's just used in a non-API way, locally by this extension, I would name it by path, i.e. `source.extensions.tracers.xray.daemon`. See https://github.com/envoyproxy/envoy/blob/master/test/fuzz/common.proto for precedent. Thanks for looking into this.

Comment (marcomagdy) [reply source/extensions/tracers/xray/daemon.proto]: @htuch here it is - https://github.com/envoyproxy/envoy/pull/9350
